tech/post/2017-11-06-Kubernetes-Taints&Tolerations.md

```diff
@@ -0,0 +1,302 @@
+  if not yet running, the pods will not be scheduled on that node.
+
+## 3. Examples
+
+### Add Pods into Master
+
+In default, Kubernetes will not allow to add the pods into ```Master```,
+If you want to add the pods into ```Master```, you can use the following command:
+
+```kubectl taint nodes --all dedicated-```
+
+### Dedicated Nodes
+
+If you want to dedicate a set of nodes for exclusive use by a particular set of users,
+you can add a taint to those nodes by using the following command:
+
+```kubectl taint nodes [nodename] dedicated=groupName:NoSchedule```
+
+### Nodes with Special Hardware
+
+If you want to keep a part of nodes have special hardware,
+for example you want to prorvide GPU for AI.
+you can add a taint to the nodes that have the Special hardware.
+
+```kubectl taint nodes [nodename] special=GPU:NoSchedule```
+
+* [Refer Pages](https://github.com/kubernetes/community/blob/58b1c30d95719749068497ba35dfe4c64b21aa72/contributors/design-proposals/scheduling/taint-toleration-dedicated.md)
+
+## 4. Taint based Evictions
+
+Taint based Evictions is an alpha feature that causes the node controller to create Taints corresponding to node condition.
+This feature is disabled in default, The users can switch on this feature by setting ```--feature-gates mapStringBool```
+in [Kubernetes Controller Manager configuration](https://kubernetes.io/docs/admin/kube-controller-manager/)
+
+* TaintBasedEvictions=true|false (ALPHA - default=false)
+
+The built-in Taints currently include the below and so forth:
+
+* node.alpha.kubernetes.io/unreachable
+
+  Current node is unreachable from the node controller.
+
+  ```Ready==Unknown```
+
+* node.alpha.kubernetes.io/notReady
+
+  Current node status is not ready.
+
+  ```Ready==False```
+
+* node.kubernetes.io/memoryPressure
+
+  Current node has memory pressure.
+
+* node.kubernetes.io/diskPressure
+
+  Current node has memory pressure.
+
+* node.kubernetes.io/network-unavailable
+
+  Current node’s network is unavailable.
+
+### Default Pods
+
+```node.alpha.kubernetes.io/notReady``` and ```node.alpha.kubernetes.io/unreachable``` will be added into the pods by default.
+
+[Added Codes](https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/admission/defaulttolerationseconds/admission.go#L66)
+
+Tolerations for Default Pods:
+
+```yaml
+
+spec:
+  containers:
+  - image: nginx
+    imagePullPolicy: Always
+    name: nginx
+  tolerations:
+  - effect: NoExecute
+    key: node.alpha.kubernetes.io/notReady
+    operator: Exists
+    tolerationSeconds: 300
+  - effect: NoExecute
+   key: node.alpha.kubernetes.io/unreachable
+    operator: Exists
+    tolerationSeconds: 300
+
+```
+
+This prevents massive pod evictions in scenarios such as the master becoming partitioned from the nodes.
+These automatically-added tolerations ensure that the default pod behavior of remaining bound for 5 minutes
+after one of these problems is detected is maintained.

```

Comment:
one of these problems detected is maintained?

Reply:
Done. Delete the words 'is maintained'